### PR TITLE
Tied the `HyperdriveDeployerCoordinator` to factory instances

### PR DESCRIPTION
**Pull request by jalextowle**
# Resolved Issues

Spearbit Issue [#15](https://github.com/spearbit-audits/delv-review/issues/15)

# Description

In order to fix the critical issue from the most recent Spearbit audit, this PR associates each deployer coordinator with a factory instance. This factory instance is the only contract that is permitted to call the deployer coordinator, which mitigates the issue that allows an approval to be drained by a malicious actor.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity and/or Rust, the corresponding checklist can
be removed.

### Solidity

- [x] **Tokens**
    - [x] Do all `approve` calls use `forceApprove`?
    - [x] Do all `transfer` calls use `safeTransfer`?
    - [x] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [x] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [x] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [x] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [x] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [x] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [x] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Rust

- [x] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### Solidity

- [x] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [x] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Rust

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### CODE REVIEW

File: contracts/src/interfaces/IHyperdriveDeployerCoordinator.sol [hunk 1]
```diff
@@ -59,6 +59,10 @@ interface IHyperdriveDeployerCoordinator {
     ///         duration or is not a multiple of the checkpoint duration.
     error InvalidPositionDuration();
 
+    /// @notice Thrown when the sender of a `deploy`, `deployTarget`, or
+    ///         `initialize` transaction isn't the associated factory.
+    error InvalidSender();
```
Comment (jrhea) [inline contracts/src/interfaces/IHyperdriveDeployerCoordinator.sol:64]: If this is only used in the `onlyFactory` modifier, then should we make the name more specific?  The comment describes a much, much more specific use for the error than the name of the error implies.

File: contracts/src/deployers/HyperdriveDeployerCoordinator.sol [hunk 2]
```diff
@@ -155,16 +171,17 @@ abstract contract HyperdriveDeployerCoordinator is
...
             deployment.target4,
-            // NOTE: We hash the sender and deployment ID with the salt to
-            // prevent the front-running of deployments.
-            keccak256(abi.encode(msg.sender, deploymentId, salt))
+            // NOTE: We hash the deployment ID with the salt to prevent the
+            // front-running of deployments.
+            keccak256(abi.encode(deploymentId, salt))
         );
-        _deployments[msg.sender][_deploymentId].hyperdrive = hyperdrive;
+        _deployments[_deploymentId].hyperdrive = hyperdrive;
```
Comment (jrhea) [inline contracts/src/deployers/HyperdriveDeployerCoordinator.sol:178]: shouldn't you just do:

`deployment.hyperdrive = hyperdrive;`

File: contracts/src/deployers/HyperdriveDeployerCoordinator.sol [hunk 3]
```diff
@@ -216,25 +231,23 @@ abstract contract HyperdriveDeployerCoordinator is
...
             );
 
             // Store the deployment.
-            _deployments[msg.sender][_deploymentId].configHash = configHash_;
-            _deployments[msg.sender][_deploymentId]
-                .extraDataHash = extraDataHash;
-            _deployments[msg.sender][_deploymentId]
-                .initialSharePrice = initialSharePrice;
-            _deployments[msg.sender][_deploymentId].target0 = target;
+            _deployments[_deploymentId].configHash = configHash_;
```
Comment (jrhea) [inline contracts/src/deployers/HyperdriveDeployerCoordinator.sol:240]: does it make sense to use a storage pointer in this function?

File: contracts/src/deployers/HyperdriveDeployerCoordinator.sol [hunk 4]
```diff
@@ -216,25 +231,23 @@ abstract contract HyperdriveDeployerCoordinator is
...
+            _deployments[_deploymentId].extraDataHash = extraDataHash;
+            _deployments[_deploymentId].initialSharePrice = initialSharePrice;
+            _deployments[_deploymentId].target0 = target;
```
Comment (jrhea) [inline contracts/src/deployers/HyperdriveDeployerCoordinator.sol:243]: does it make sense to use a storage pointer for _deployments[_deploymentId] in this function?

**Review by jrhea — APPROVED** [review]
a couple of questions; otherwise, lgtm

**Review by mcclurejt — APPROVED** [review]

**Pull request merged**